diff --git a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
index 06102247c0f5..12806e64da43 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
@@ -1836,7 +1836,7 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
   dpy = gst_va_display_get_va_dpy (self->display);
 
   sample = tx->next (tx->user_data);
-  for (; sample; sample = tx->next (tx->user_data)) {
+  for(int i=0; sample; i++, sample = tx->next (tx->user_data)) {
     VAProcPipelineParameterBuffer params = { 0, };
     VABufferID buffer;
     VASurfaceID in_surface;
@@ -1879,6 +1879,15 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       return FALSE;
     }
 
+    if(0==i)
+    {//With alpha, set black background color
+       params.output_background_color = 0xff000000;
+    }
+    else
+    {//Without alpha, transparent background
+       params.output_background_color = 0;
+    }
+
     status = vaCreateBuffer (dpy, self->context,
         VAProcPipelineParameterBufferType, sizeof (params), 1, &params,
         &buffer);
